src/CHIPBackend.hh

```diff
@@ -639,9 +636,9 @@ public:
   CHIPEventFlags getFlags() { return Flags_; }
   std::mutex EventMtx;
   std::string Msg;
-  size_t getCHIPRefc();
-  virtual void decreaseRefCount(std::string Reason);
-  virtual void increaseRefCount(std::string Reason);
+  virtual size_t getCHIPRefc() = 0;
```

Comment:
To me, moving these down to Level0 BE seems sensible as OpenCL BE does not need to implement these.

Reply:
but we can use native OpenCL refcount tracking to verify that ours is working correctly, plus we want to have as much code as possible in the shared code paths - the backend implementations should be as thin as possible and use CHIP functions when possible.

Reply:
> but we can use native OpenCL refcount tracking to verify that ours is working correctly

If by that you mean: comparing the numbers coming from `clGetEventInfo(CL_EVENT_REFERENCE_COUNT)` with a separate refcount kept in `CHIPEvent` - that does not work in general. See e.g. [official OpenCL documentation](https://github.com/KhronosGroup/OpenCL-Docs/issues/620) on object lifetime, in particular:

> Implementations may also modify the reference count, e.g. to track attached objects or to ensure correct operation of in-progress or scheduled activities

PoCL does this in multiple places, e.g. with every `clEnqueueNDRangeKernel` we retain the `cl_kernel` plus every `cl_mem` object used to set kernel's arguments. This is necessary because the user is permitted to call `clReleaseKernel` right after enqueuing the kernel, and then use `clFinish` on the queue. I'd be surprised if other implementations don't do this.

Reply:
Ah I see, the OpenCL refcount can change in unpredictable ways depending on the OpenCL implementation. You can mark all comments about losing OpenCL refcounting as resolved, but I think that `increaseRefCount()` and related functionality should still be kept at the CHIP-SPV level if not for any other reason than to keep backend implementations as thin as possible.